src/server/worker/api_server.go

```diff
@@ -457,23 +467,35 @@ func (a *APIServer) runUserCode(ctx context.Context, logger *taggedLogger, envir
 		},
 	}
 	cmd.Dir = a.workingDir
-	err := cmd.Run()
-
-	// Return result
-	if err == nil {
-		return nil
+	err := cmd.Start()
+	if err != nil {
+		return err
 	}
-	// (if err is an acceptable return code, don't return err)
-	if exiterr, ok := err.(*exec.ExitError); ok {
-		if status, ok := exiterr.Sys().(syscall.WaitStatus); ok {
-			for _, returnCode := range a.pipelineInfo.Transform.AcceptReturnCode {
-				if int(returnCode) == status.ExitStatus() {
-					return nil
+	cmdErr := make(chan error, 1)
+	go func() {
+		cmdErr <- cmd.Wait()
+	}()
```

Comment:
I split this up to ameliorate [this issue](https://github.com/pachyderm/pachyderm/issues/2544)

There's also a known issue w `cmd.Run` where if you're doing IO not to a file descriptor (among other edge cases) cancelling the context won't halt command execution. So for now, I put the command in a separate channel so that the context timeout is respected. The zombie killing solution mentioned above will be added as another iteration.

Reply:
For posterity: the issue with `cmd.Run` is, we believe: https://github.com/golang/go/issues/18874

h/t @robert-uhl who found it